Cabal/Distribution/Parsec/Types/Field.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE DeriveFunctor #-}
+module Distribution.Parsec.Types.Field (
+    -- * Cabal file
+    Field (..),
+    FieldLine (..),
+    SectionArg (..),
+    sectionArgAnn,
+    -- * Name
+    Name (..),
+    mkName,
+    getName,
+    -- * Position
+    Position (..),
+    incPos,
+    retPos,
+    showPos,
+    ) where
+
+import Data.ByteString (ByteString)
+
+import qualified Data.Char as Char
+import qualified Data.ByteString.Char8 as B
+
+-------------------------------------------------------------------------------
+-- Cabal file
+-------------------------------------------------------------------------------
+
+data Field ann
+    = Field   !(Name ann) [FieldLine ann]
+    | Section !(Name ann) [SectionArg ann] [Field ann]
+    | IfElseBlock [SectionArg ann] [Field ann] [Field ann]
+  deriving (Eq, Show, Functor)
+
+data FieldLine ann  = FieldLine  !ann !ByteString
+  deriving (Eq, Show, Functor)
+
+data SectionArg ann
+    = SecArgName  !ann !ByteString
+    | SecArgStr   !ann !String
+    | SecArgNum   !ann !ByteString
+    | SecArgOther !ann !ByteString
+  deriving (Eq, Show, Functor)
+
+-- | Extract annotation from 'SectionArg'.
+sectionArgAnn :: SectionArg ann -> ann
+sectionArgAnn (SecArgName ann _)  = ann
+sectionArgAnn (SecArgStr ann _)   = ann
+sectionArgAnn (SecArgNum ann _)   = ann
+sectionArgAnn (SecArgOther ann _) = ann
+
+-------------------------------------------------------------------------------
+-- Name
+-------------------------------------------------------------------------------
+
+data Name ann  = Name       !ann !ByteString
+  deriving (Eq, Show, Functor)
+
+mkName :: ann -> ByteString -> Name ann
+mkName ann bs = Name ann (B.map Char.toLower bs)
+
+getName :: Name a -> ByteString
+getName (Name _ bs) = bs
+
+-------------------------------------------------------------------------------
+-- Position
+-------------------------------------------------------------------------------
+
+data Position = Position
```

Comment:
1. I'm quite sure that beginning of the token would be enough (the end of token is easy to calculate for all tokens, and also `Field`).
2. `simpleParsec` will go away, we will always have some Position.